### ISSUE TITLE
cse optimizations don't handle x**2 == x*x and 2*x == x + x

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#10228]
```
In [11]: cse([x*y, x*x*y], optimizations='basic')
Out[11]:
⎛    ⎡      2  ⎤⎞
⎝[], ⎣x⋅y, x ⋅y⎦⎠

In [13]: cse([x + y, 2*x + y], optimizations='basic')
Out[13]: ([], [x + y, 2⋅x + y])
```

We should add optimizations to split out integer powers to multiplications and integer multiples to additions. Some care must be taken to avoid unnecessary splitting for very large powers or multiples.

**Comment by smichr** [comment]
```python
>>> (x**2*y).extract_multiplicatively(x*y)
x
>>> (2*x+y).extract_additively(x+y)
x
```

**Comment by asmeurer** [comment]
I'm not sure if that generalizes. Those might both appear as subexpressions of larger expressions.

**Comment by jschall** [comment]
Darn, I tested 10271 thinking that it was a fix for this issue and figured out that it isn't. Is there a fix available for this issue?

@asmeurer

**Comment by asmeurer** [comment]
No. I believe it can be done with a optimization that converts `2*x` to `Add(x, x, evaluate=False)` (meaning you could pass a custom optimization to `cse` to make it work).

**Comment by smichr** [comment]
> I'm not sure if that generalizes. Those might both appear as subexpressions of larger expressions.

I'm pretty sure it will work regardless of the length of the self-same expression it is part of:

```python
>>> (2*x+y+z).extract_additively(x+y)
x + z
>>> (2*x**3*y**2*z).extract_multiplicatively(x**2*y)
2*x*y*z
```

**Comment by smichr** [comment]
My `cse` branch starts to address this.

**Comment by smichr** [comment]
reopen after cse mod reversion

Note: I suspect the slow down for larger expressions comes from the pairwise comparison to try find the best expressions to extract. I recall seeing something in one of the low-level routines where this issue was recognized -- trying to identify subexpressions amounts Add exponents, as I recall. I suspect the way to go with this is to have a keyword `best` to indicate that the best solution should be sought (which will, in general, be expensive) and when that is false, accept something that is "good enough". I recall hearing that factoring is probabilistic. I wonder if this might be though of as a factoring problem where you are trying to find the sets of factors among *subsets* of terms that will lead to a compact expression, like `x*y + x*a + y*z + x*a*z` -> `x(a(1+z)+y)+yz` or `a(x(z+1))+y(x+z)`.

**Comment by asmeurer** [comment]
Perhaps relevant: https://github.com/sympy/sympy/pull/11586 (and https://github.com/sympy/sympy/issues/11349).
